packages/angular_devkit/build_angular/src/builders/web-test-runner/index.ts

```diff
@@ -0,0 +1,199 @@
+
+    return testOutput;
+  },
+);
+
+/** Build all the given test files and write the result to the given output path. */
+async function buildTests(
+  testFiles: string[],
+  outputPath: string,
+  options: WtrBuilderOptions,
+  ctx: BuilderContext,
+): Promise<BuilderOutput> {
+  const jasmine = path.relative(
+    process.cwd(),
+    require.resolve('jasmine-core/lib/jasmine-core/jasmine.js'),
+  );
+  const jasmineRunner = path.relative(process.cwd(), path.join(__dirname, 'jasmine_runner.js'));
+
+  // Build tests with `application` builder, using test files as entry points.
+  // Also bundle in Jasmine and the Jasmine runner script, which need to share chunked dependencies.
+  const buildOutput = await first(
+    buildApplicationInternal(
+      {
+        entryPoints: new Set([...testFiles, jasmine, jasmineRunner]),
+        tsConfig: options.tsConfig,
+        outputPath,
+        aot: false,
+        index: false,
+        outputHashing: OutputHashing.None,
+        optimization: false,
+        externalDependencies: [
+          // Resolved by `@web/test-runner` at runtime with dynamically generated code.
+          '@web/test-runner-core',
+        ],
+        sourceMap: {
+          scripts: true,
+          styles: true,
+          vendor: true,
+        },
+        polyfills: withZonePolyfills(options.polyfills),
+      },
+      ctx,
+    ),
+  );
+
+  return buildOutput;
+}
+
+/** Run Web Test Runner on the given directory of bundled JavaScript tests. */
+async function runTests(
+  wtr: typeof WebTestRunner,
+  testDir: string,
+  options: WtrBuilderOptions,
+): Promise<BuilderOutput> {
+  const testPagePath = path.resolve(__dirname, 'test_page.html');
+  const testPage = await fs.readFile(testPagePath, 'utf8');
+
+  const runner = await wtr.startTestRunner({
+    config: {
+      rootDir: testDir,
+      files: [
+        `${testDir}/**/*.js`,
+        `!${testDir}/polyfills.js`,
+        `!${testDir}/chunk-*.js`,
+        `!${testDir}/jasmine.js`,
+        `!${testDir}/jasmine_runner.js`,
+      ],
+      testFramework: {
+        config: {
+          defaultTimeoutInterval: 5_000,
+        },
+      },
+      nodeResolve: true,
+      port: 9876,
+      watch: options.watch ?? false,
+
+      testRunnerHtml: (_testFramework, _config) => testPage,
+    },
+    readCliArgs: false,
+    readFileConfig: false,
+    autoExitProcess: false,
+  });
+  if (!runner) {
+    throw new Error('Failed to start Web Test Runner.');
+  }
+
+  // Wait for the tests to complete and stop the runner.
+  const passed = (await once(runner, 'finished')) as boolean;
+  await runner.stop();
+
+  // No need to return error messages because Web Test Runner already printed them to the console.
+  return { success: passed };
+}
+
+/** Returns a list of polyfills copied from the given polyfills but with `zone.js` and `zone.js/testing` added if necessary. */
+function withZonePolyfills(inputPolyfills: string[]): string[] {
```

Comment:
I don't think we should add polyfills when users did not add them explicitly.

Reply:
I dropped `withZonePolyfills` altogether, so we should just directly respect the user's provided `polyfills` array.

You pointed me at [Karma's implementation](https://github.com/angular/angular-cli/blob/7a50df5c046b65bb196408dd2c4771d5deccd034/packages/angular_devkit/build_angular/src/tools/webpack/configs/common.ts#L133-L161) separately and it feels like this is effectively moving a `zone.js/testing` polyfill to the end of the start of the main test bundle.

I updated this PR to make a new entry point for `zone.js/testing` and load that after Jasmine. I believe that should align with Karma's file ordering behavior. I had to make a small edit to `application` builder to allow bare module specifiers in entry points because `require.resolve('zone.js/testing')` was giving me the UMD module. Side effect of that is that I can take your original suggestion in the other comment for using `jasmine-core/...` and `.../jasmine_runner.js` as entry points directly and drop their `require.resolve()` calls.

I need to double check that this doesn't break Jest along the way, since that was the existing usage of this, but otherwise I think this looks good.